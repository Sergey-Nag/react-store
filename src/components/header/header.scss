$avatar-size: 40px;

.avatar {
  width: $avatar-size;
  height: $avatar-size;
  border-radius: 50%;
  background-color: var(--bs-gray-300);
  border: 1px solid var(--bs-gray-400);
  display: flex;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
  font-weight: 500;
  color: var(--bs-gray-600);
}