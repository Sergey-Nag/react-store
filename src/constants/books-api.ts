export const BOOKS_API_BASE_URL = 'https://js-band-store-api.glitch.me';