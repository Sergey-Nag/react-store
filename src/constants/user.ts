export const USER_LOCAL_STATE_KEY = 'user-data';