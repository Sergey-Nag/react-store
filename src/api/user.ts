import { BOOKS_API_BASE_URL } from "../constants/books-api";
import { User } from "../types/user";
import { request } from "./reqest";

export const loginUser = async (username: string | null): Promise<User> => 
  await request<User>(`${BOOKS_API_BASE_URL}/signin`, 'POST', { username });
