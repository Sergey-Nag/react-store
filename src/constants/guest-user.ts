import { User } from "../types/user";

export const GUEST_USER = {
  username: 'Guest',
} as User;