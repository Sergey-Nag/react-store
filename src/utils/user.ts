import { USER_LOCAL_STATE_KEY } from "../constants/user";
import { User } from "../types/user";

export const getLocalStorageUser = (): User | null => {
  const user = localStorage.getItem(USER_LOCAL_STATE_KEY);

  return user ? JSON.parse(user) : null;
};

export const setLocalStorageUser = (user: User) =>
  localStorage.setItem(USER_LOCAL_STATE_KEY, JSON.stringify(user));

export const clearLocalStorageUser = () =>
  localStorage.removeItem(USER_LOCAL_STATE_KEY);
