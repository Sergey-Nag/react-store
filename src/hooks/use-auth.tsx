import { createContext, PropsWithChildren, useContext, useState } from "react"
import { loginUser } from "../api/user";
import { User } from "../types/user";
import { clearLocalStorageUser, getLocalStorageUser, setLocalStorageUser } from "../utils/user";

interface AuthState {
  user: User | null,
  login: (username: string) => Promise<User | null>,
  logout: () => void,
  authError: Error | null
}

const initialAuthState: AuthState = {
  user: null,
  authError: null,
  login() { return Promise.resolve(null) },
  logout() {},
};

const AuthContext = createContext(initialAuthState);

export function AuthProvider(props: PropsWithChildren) {
  const [userData, setUserData] = useState<User | null>(getLocalStorageUser);
  const [authError, setAuthError] = useState<Error | null>(null);

  const login = async (username: string) => {
    if (userData) return null;
    let userResponse;
    try {
      userResponse = await loginUser(username);
      setUserData(userResponse);
      setLocalStorageUser(userResponse);
      setAuthError(null);
    } catch(e: any) {
      setAuthError(e);
    }

    return userResponse ?? null;
  };

  const logout = () => {
    clearLocalStorageUser();
    setUserData(null);
  };

  const authContextValue: AuthState = { 
    user: userData,
    authError,
    login,
    logout,
  }

  return <AuthContext.Provider value={authContextValue} {...props} />
}

export const useAuth = () => useContext(AuthContext);