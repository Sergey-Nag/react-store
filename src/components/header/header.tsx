import { useMemo } from "react";
import { GUEST_USER } from "../../constants/guest-user";
import { useAuth } from "../../hooks/use-auth";
import { DropdownButton } from "../dropdown-button/dropdown-button";
import './header.scss';

export function Header() {
  const { user, logout } = useAuth();
  const { username } = useMemo(() => user ?? GUEST_USER , [user]);

  const getUsernamePlaceholder = (username: string) => username.slice(0, 2);

  return (
    <nav className="navbar navbar-expand-lg navbar-light bg-light">
      <div className="container justify-content-end">
      <div className="d-flex align-items-baseline">
        <div className="avatar me-2">
          { getUsernamePlaceholder(username) }
        </div>
        <span className="fw-bold">{ username }</span>
        { user && 
          <DropdownButton className="ms-2">
            <button
              className="btn"
              onClick={logout}
            >
              Logout
            </button>
          </DropdownButton>
        }
      </div>
      </div>
    </nav>
  );
}