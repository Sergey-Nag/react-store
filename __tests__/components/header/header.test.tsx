import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from '../../../src/components/header/header';
import { USER_LOCAL_STATE_KEY } from '../../../src/constants/user';
import { AuthProvider, useAuth } from '../../../src/hooks/use-auth';
import { User } from '../../../src/types/user';

const mockAuthState: {
  user: User | null,
  login: jest.Mock,
  logout: jest.Mock,
} = {
  user: null,
  login: jest.fn(() => Promise.resolve()),
  logout: jest.fn()
}

const renderWithAuthProvider = (authState = mockAuthState) =>
  render(
    <AuthProvider value={authState}>
      <Header />
    </AuthProvider>
  );

describe('Header' , () => {
  afterEach(() => {
    localStorage.clear();
  });

  it('Should render component with username and dropdown if user exists', () => {
    localStorage.setItem(
      USER_LOCAL_STATE_KEY, 
      JSON.stringify({
        avatar: 'test',
        username: 'test@test.t',
        token: 'test'
      })
    );

    const { asFragment } = renderWithAuthProvider();

    expect(asFragment()).toMatchSnapshot('with username');
  });

  it('Should render component with guest username and without dropdown if user doesn\'t exist', () => {
    const { asFragment } = renderWithAuthProvider();

    expect(asFragment()).toMatchSnapshot('with guest username');
  });

  it('Should logout current user clicking by logout button in the dropdown', () => {
    localStorage.setItem(
      USER_LOCAL_STATE_KEY, 
      JSON.stringify({
        avatar: 'test',
        username: 'test@test.t',
        token: 'test'
      })
    );

    renderWithAuthProvider();
    console.log(mockAuthState);
    
    fireEvent.click(screen.getByRole('menu'));
    fireEvent.click(screen.getByRole('menuitem'), { name: 'Logout'});
    
    expect(mockAuthState.logout).toHaveBeenCalled();
    // expect(mockAuthState.user).toBeNull();
  });
});