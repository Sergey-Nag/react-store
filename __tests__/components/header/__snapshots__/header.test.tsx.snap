// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Should render component with guest username and without dropdown if user doesn't exist: with guest username 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <div
      class="container justify-content-end"
    >
      <div
        class="d-flex align-items-baseline"
      >
        <div
          class="avatar me-2"
        >
          Gu
        </div>
        <span
          class="fw-bold"
        >
          Guest
        </span>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Header Should render component with username and dropdown if user exists: with username 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <div
      class="container justify-content-end"
    >
      <div
        class="d-flex align-items-baseline"
      >
        <div
          class="avatar me-2"
        >
          te
        </div>
        <span
          class="fw-bold"
        >
          test@test.t
        </span>
        <div
          class="position-relative"
          role="menu"
        >
          <button
            class="ms-2 btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <div
            class="dropdown-menu dropdown-right"
          >
            <button
              class="btn dropdown-item"
              role="menuitem"
            >
              Logout
            </button>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
